js/runner.go

```diff
@@ -363,57 +362,46 @@ type VU struct {
 	TLSConfig *tls.Config
 	ID        int64
 	Iteration int64
+	runMutex  *sync.Mutex
 
 	Console *console
 	BPool   *bpool.BufferPool
 
 	Samples chan<- stats.SampleContainer
 
 	setupData goja.Value
-
-	// A VU will track the last context it was called with for cancellation.
-	// Note that interruptTrackedCtx is the context that is currently being tracked, while
-	// interruptCancel cancels an unrelated context that terminates the tracking goroutine
-	// without triggering an interrupt (for if the context changes).
-	// There are cleaner ways of handling the interruption problem, but this is a hot path that
-	// needs to be called thousands of times per second, which rules out anything that spawns a
-	// goroutine per call.
-	interruptTrackedCtx context.Context
-	interruptCancel     context.CancelFunc
-
-	m *sync.Mutex
 }
 
-// Verify that VU implements lib.VU
-var _ lib.VU = &VU{}
+// Verify that interfaces are implemented
+var _ lib.ActiveVU = &ActiveVU{}
+var _ lib.InitializedVU = &VU{}
 
-func (u *VU) Reconfigure(id int64) error {
-	u.ID = id
-	u.Iteration = 0
-	u.Runtime.Set("__VU", u.ID)
-	return nil
+// ActiveVU holds a VU and its activation parameters
+type ActiveVU struct {
+	*VU
+	*lib.VUActivationParams
 }
 
-func (u *VU) RunOnce(ctx context.Context) error {
-	u.m.Lock()
-	defer u.m.Unlock()
-	// Track the context and interrupt JS execution if it's cancelled.
-	if u.interruptTrackedCtx != ctx {
-		interCtx, interCancel := context.WithCancel(context.Background())
-		if u.interruptCancel != nil {
-			u.interruptCancel()
+// Activate the VU so it will be able to run code
+func (u *VU) Activate(params *lib.VUActivationParams) lib.ActiveVU {
+	u.Runtime.ClearInterrupt()
+	// u.Env = params.Env
+
+	go func() {
+		<-params.RunContext.Done()
+		u.Runtime.Interrupt(errInterrupt)
+		if params.DeactivateCallback != nil {
```

Comment:
Shouldn't we surround this `if` with `u.runMutex.Lock()` and `u.runMutex.Unlock()`. After all, we don't want to return the VU until its execution has finished. And I still think a mutex leaves a minuscule chance of a race, but we can fix that in a later PR.

Reply:
Even that locking won't be enough to prevent all data races, as demonstrated by https://github.com/loadimpact/k6/issues/1404. But, as I mention in the issue, I think we can merge this for now, and fix it properly later. It's probably a better state than the current bug :sweat_smile: